ignite/metrics/ssim.py

```diff
@@ -0,0 +1,178 @@
+from typing import Callable, Sequence
+
+import torch
+import torch.nn.functional as F
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["SSIM"]
+
+
+class SSIM(Metric):
+    """
+    Computes Structual Similarity Index Measure
+
+    Args:
+        kernel_size (list or tuple of int): Size of the gaussian kernel. Default: (11, 11)
+        sigma: Standard deviation of the gaussian kernel. Default: (1.5, 1.5)
+        data_range: Range of the image. If ``None``, it is determined from the image (max - min)
+        k1: Parameter of SSIM. Default: 0.01
+        k2: Parameter of SSIM. Default: 0.03
+        output_transform: A callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric.
+
+    Returns:
+        A Tensor with SSIM
+
+    Example:
+
+    To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+    The output of the engine's ``process_function`` needs to be in the format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        metric = SSIM()
+        metric.attach(engine, "ssim")
+
+    If the output of the engine is not in the format above, ``output_transform`` argument can be used to transform it.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return {'prediction': y_pred, 'target': y, ...}
+
+        engine = Engine(process_function)
+
+        def output_transform(output):
+            # `output` variable is returned by above `process_function`
+            y_pred = output['prediction']
+            y = output['target']
+            return y_pred, y  # output format is according to `Accuracy` docs
+
+        metric = SSIM(output_transform=output_transform)
+        metric.attach(engine, "ssim")
+
+    The user even can use the metric with ``update`` and ``compute`` methods.
+
+    .. code-block:: python
+
+        >>> y_pred = torch.rand([16, 1, 16, 16])
+        >>> y = y_pred * 1.25
+        >>> ssim = SSIM()
+        >>> ssim.update((y_pred, y))
+        >>> ssim.compute()
+        tensor(0.9520)
+    """
+
+    def __init__(
+        self,
+        kernel_size: Sequence[int] = (11, 11),
+        sigma: Sequence[float] = (1.5, 1.5),
+        data_range: float = None,
+        k1: float = 0.01,
+        k2: float = 0.03,
+        output_transform: Callable = lambda x: x,
+    ):
+        if len(kernel_size) != 2 or len(sigma) != 2:
```

Comment:
Maybe, we should also check the type of input args. For example, user can set `kernel_size` as a scalar (which can be also natural), so `len(kernel_size)` will fail with an unrelated error.

Reply:
I set `kernel_size` and `sigma` to be a list or tuple of length two to allow the users define different width or height of `kernel_size`. I know the user will want to set `kernel_size` as a scalar if the width and height are going to be same. For this, we might need another checking. That's why I let the users define in a concise way for the same or different width and height with minimal internal working code.

What do you think?

Reply:
Yes, I understand your choice. Ideally, it would be nice to support both types, like pytorch does for convolutions and pooling ops

Reply:
Alright, I will add them